### PR TITLE
fix: Not generate chunks with the same content

### PR DESCRIPTION
**Pull request by mshabarov**
## Description

Checks if a chunk has already been considered for a given list of imported modules and scripts.
If so, it doesn't generate a new file, but only updates the the import file, linking all the route class hashes with the same chunk file.

Also adds `dependencyTriggers` and `eager` fields to the `ChunkInfo` hashCode and equals, as `ChunkInfo` is used widely in maps.

Adds extra debug logging to `AbstractUpdateImports` to understand better what takes most of the time in large projects.

Related-to https://github.com/vaadin/flow/issues/17234

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/AbstractUpdateImports.java [hunk 1]
```diff
@@ -188,42 +222,68 @@ private Map<File, List<String>> process(Map<ChunkInfo, List<CssData>> css,
...
+                String chunkFilename = "chunk-" + chunkContentHash + ".js";
+
+                String ifClauses = chunkInfo.getDependencyTriggers().stream()
+                        .map(BundleUtils::getChunkId)
+                        .map(hash -> "key === '" + hash + "'")
+                        .collect(Collectors.joining(" || "));
+                chunkLoader.add("  if (" + ifClauses + ") {");
+                chunkLoader.add("    pending.push(import('./chunks/"
+                        + chunkFilename + "'));");
+                chunkLoader.add("  }");
```
Comment (caalador) [inline flow-server/src/main/java/com/vaadin/flow/server/frontend/AbstractUpdateImports.java:271]: Just as a simpler read of the resulting strings I would suggest using String.format("");
```
                String ifClauses = chunkInfo.getDependencyTriggers().stream()
                        .map(BundleUtils::getChunkId)
                        .map(hash -> String.format("key === '%s'", hash))
                        .collect(Collectors.joining(" || "));
                chunkLoader.add(String.format("  if (%s) {", ifClauses));
                chunkLoader.add(String.fotmat("    pending.push(import('./chunks/%s'));",
                        chunkFilename);
                chunkLoader.add("  }");
```

Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/AbstractUpdateImports.java:271]: done

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/AbstractUpdateImports.java [hunk 2]
```diff
@@ -188,42 +222,68 @@ private Map<File, List<String>> process(Map<ChunkInfo, List<CssData>> css,
...
+                    files.put(chunkFile, chunkLines);
+                }
             }
 
             chunkLoader.add("  return Promise.all(pending);");
             chunkLoader.add("}");
             chunkLoader.add("");
+
+            ms = (System.nanoTime() - start) / 1000000;
+            getLogger().debug("Lazy chunks generation took {} ms.", ms);
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/AbstractUpdateImports.java:286]: done

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/AbstractUpdateImports.java [hunk 3]
```diff
@@ -117,18 +117,49 @@ abstract class AbstractUpdateImports implements Runnable {
...
+        long start = System.nanoTime();
+
         Map<ChunkInfo, List<CssData>> css = scanner.getCss();
+        Map<ChunkInfo, List<String>> javascript = getMergedJavascript();
+
+        Map<File, List<String>> output = process(css, javascript);
+        writeOutput(output);
+
+        long ms = (System.nanoTime() - start) / 1000000;
+        getLogger().debug("Imports and chunks update took {} ms.", ms);
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/AbstractUpdateImports.java]: done

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/AbstractUpdateImports.java [hunk 4]
```diff
@@ -117,18 +117,49 @@ abstract class AbstractUpdateImports implements Runnable {
...
+
         Map<ChunkInfo, List<String>> javascript;
+        Map<ChunkInfo, List<String>> modules = scanner.getModules();
+        Map<ChunkInfo, List<String>> scripts = scanner.getScripts();
+
         if (options.isProductionMode()) {
-            javascript = mergeJavascript(scanner.getModules(),
-                    scanner.getScripts());
+            getLogger().debug("Found {} modules, and {} scripts.", modules,
+                    scripts);
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/AbstractUpdateImports.java]: good catch! done

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/AbstractUpdateImports.java [hunk 5]
```diff
@@ -117,18 +117,49 @@ abstract class AbstractUpdateImports implements Runnable {
...
-                    scanner.getModulesDevelopment(), scanner.getScripts(),
-                    scanner.getScriptsDevelopment());
+            Map<ChunkInfo, List<String>> modulesDevelopment = scanner
+                    .getModulesDevelopment();
+            Map<ChunkInfo, List<String>> scriptsDevelopment = scanner
+                    .getScriptsDevelopment();
+
+            getLogger().debug(
+                    "Found {} modules, {} scripts, {} devmode modules and {} devmode scripts.",
+                    modules, scripts, modulesDevelopment, scriptsDevelopment);
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/AbstractUpdateImports.java]: good catch! done

**Review by caalador — APPROVED** [review]

**Pull request merged**